Modules/_abc.c

```diff
@@ -0,0 +1,639 @@
+   In particular use capitals like PyList_GET_SIZE.
+   Think (ask) about inlining some calls, like __subclasses__.
+   Fix refleaks. */
+
+#include "Python.h"
+#include "structmember.h"
+
+PyDoc_STRVAR(_abc__doc__,
+"Module contains faster C implementation of abc.ABCMeta");
+#define DEFERRED_ADDRESS(ADDR) 0
+
+_Py_IDENTIFIER(stdout);
+_Py_IDENTIFIER(__isabstractmethod__);
+_Py_IDENTIFIER(__abstractmethods__);
+_Py_IDENTIFIER(__class__);
+
+
+/* A global counter that is incremented each time a class is
+   registered as a virtual subclass of anything.  It forces the
+   negative cache to be cleared before its next use.
+   Note: this counter is private. Use `abc.get_cache_token()` for
+   external code. */
+static Py_ssize_t abc_invalidation_counter = 0;
+
+typedef struct {
+    PyHeapTypeObject tp;
+    PyObject *abc_registry; /* these three are normal set of weakrefs without callback */
+    PyObject *abc_cache;
+    PyObject *abc_negative_cache;
+    Py_ssize_t abc_negative_cache_version;
+} abc;
+
+static void
+abcmeta_dealloc(abc *tp)
+{
+    Py_CLEAR(tp->abc_registry);
+    Py_CLEAR(tp->abc_cache);
+    Py_CLEAR(tp->abc_negative_cache);
+    Py_TYPE(tp)->tp_free((PyObject *)tp);
+}
+
+static int
+abcmeta_traverse(PyObject *self, visitproc visit, void *arg)
+{
+    Py_VISIT(((abc *)self)->abc_registry);
+    Py_VISIT(((abc *)self)->abc_cache);
+    Py_VISIT(((abc *)self)->abc_negative_cache);
+    return PyType_Type.tp_traverse(self, visit, arg);
+}
+
+static int
+abcmeta_clear(abc *tp)
+{
+    if (tp->abc_registry) {
+        PySet_Clear(tp->abc_registry);
+    }
+    if (tp->abc_cache) {
+        PySet_Clear(tp->abc_cache);
+    }
+    if (tp->abc_negative_cache) {
+        PySet_Clear(tp->abc_negative_cache);
+    }
+    return PyType_Type.tp_clear((PyObject *)tp);
+}
+
+static PyObject *
+abcmeta_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+    abc *result = NULL;
+    PyObject *ns, *bases, *items, *abstracts, *is_abstract, *base_abstracts;
+    PyObject *key, *value, *item, *iter;
+    Py_ssize_t pos = 0;
+    result = (abc *)PyType_Type.tp_new(type, args, kwds);
+    if (!result) {
+        return NULL;
+    }
+    /* Set up inheritance registry. */
+    result->abc_registry = PySet_New(NULL);
+    result->abc_cache = PySet_New(NULL);
+    result->abc_negative_cache = PySet_New(NULL);
+    if (!result->abc_registry || !result->abc_cache || !result->abc_negative_cache) {
+        Py_DECREF(result);
+        return NULL;
+    }
+    result->abc_negative_cache_version = abc_invalidation_counter;
+    if (!(abstracts = PyFrozenSet_New(NULL))) {
+        Py_DECREF(result);
+        return NULL;
+    }
+    /* Compute set of abstract method names in two stages:
+       Stage 1: direct abstract methods.
+       (It is safe to assume everything is fine since type.__new__ succeeded.) */
+    ns = PyTuple_GET_ITEM(args, 2);
+    items = PyMapping_Items(ns); /* TODO: Fast path for exact dicts with PyDict_Next */
```

Comment:
This could be null if ns is a subclass of dict that overwrites the items method to raise an error or return a non-iterable